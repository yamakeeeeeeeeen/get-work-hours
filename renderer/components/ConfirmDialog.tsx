import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  body: string;
  callback: () => void;
};

export const ConfirmDialog: FC<Props> = memo(({ isOpen, setClose, body, callback }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{body}</Typography>
        <Box display="flex">
          <Button variant="contained" color="primary" onClick={callback}>
            Yes
          </Button>
          <Button variant="contained" color="secondary" onClick={setClose}>
            No
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ConfirmDialog.displayName = 'ConfirmDialog';
