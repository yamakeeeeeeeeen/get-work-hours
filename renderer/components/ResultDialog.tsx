import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  result: { breakTime: string; uptime: string };
};

export const ResultDialog: FC<Props> = memo(({ isOpen, setClose, result }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{result.uptime}</Typography>
        <Typography variant="h4">{result.breakTime}</Typography>
        <Box display="flex">
          <Button variant="contained" color="secondary" onClick={setClose}>
            CLOSE
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ResultDialog.displayName = 'ResultDialog';
