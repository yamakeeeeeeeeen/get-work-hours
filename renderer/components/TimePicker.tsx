import { createStyles, makeStyles, TextField, Theme } from '@material-ui/core';
import { FC } from 'react';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 120,
    },
  }),
);

type Props = {
  label: string;
};

export const TimePicker: FC<Props> = ({ label }) => {
  const classes = useStyles();

  return (
    <TextField
      label={label}
      type="time"
      defaultValue="12:00"
      variant="outlined"
      color="primary"
      className={classes.textField}
      InputLabelProps={{
        shrink: true,
      }}
    />
  );
};
