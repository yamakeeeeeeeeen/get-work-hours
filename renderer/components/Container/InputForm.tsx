import { FC, useCallback, useMemo, useState } from 'react';
import { useFormContext } from 'react-hook-form';
import { Component } from '~/components/Presenter/InputForm';
import { Append, Fields, Remove } from '~/@types/ReactHookForm';
import { Inputs } from '~/pages';
import useBooleanState from '~/hooks/useBooleanState';
import { getUptimeAndBreakTime } from '~/utils/getUptimeAndBreakTime';
import { calcTime } from '~/utils/calcTime';

type Props = {
  fields: Fields;
  append: Append;
  remove: Remove;
};

type Time = { hour: number; minute: number };
const INITIAL_TIME: Time = {
  hour: 0,
  minute: 0,
};

export const InputForm: FC<Props> = ({ fields, append, remove }) => {
  const { getValues, handleSubmit } = useFormContext();
  const [isOpen, setOpen, setClose] = useBooleanState(false);
  const [uptime, setUptime] = useState<Time>(INITIAL_TIME);
  const [breakTime, setBreakTime] = useState<Time>(INITIAL_TIME);

  const result = useMemo(
    () => `
    稼働時間は${uptime.hour}時間${uptime.minute}分です。
    休憩時間は${breakTime.hour}時間${breakTime.minute}分です。
  `,
    [breakTime.hour, breakTime.minute, uptime.hour, uptime.minute],
  );

  const handleAppend = useCallback(() => {
    const values = getValues();
    const workTimes = values.workTimes;
    const lastWorkTime = workTimes[workTimes.length - 1];
    const nextTime = calcTime(lastWorkTime.start, lastWorkTime.end);
    console.log('nextTime', nextTime);
    append({ start: '12:00', end: '12:00' });
  }, [append, getValues]);
  const handleConfirm = useCallback(
    (val: Inputs) => {
      const workTimes = val.workTimes;
      const data = getUptimeAndBreakTime(workTimes);
      setUptime(data.uptime);
      setBreakTime(data.breakTime);
      setOpen();
    },
    [setOpen],
  );

  return (
    <Component
      fields={fields}
      remove={remove}
      result={result}
      isOpen={isOpen}
      setClose={setClose}
      handleAppend={handleAppend}
      handleConfirm={handleConfirm}
      handleSubmit={handleSubmit}
    />
  );
};
