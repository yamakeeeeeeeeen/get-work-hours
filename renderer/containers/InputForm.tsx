import { FC, memo, useCallback, useMemo, useState } from 'react';
import { useFormContext } from 'react-hook-form';
import { Component } from '~/components/InputForm';
import { ComponentProps, Inputs } from '~/pages';
import useBooleanState from '~/hooks/useBooleanState';
import { getWorkingTime } from '~/utils/getWorkingTime';

export type Props = ComponentProps;

type Time = { hour: number; minute: number };
const INITIAL_TIME: Time = {
  hour: 0,
  minute: 0,
};

export const InputForm: FC<Props> = memo(({ append, ...props }) => {
  const { formState, handleSubmit, reset } = useFormContext();
  const { isDirty } = formState;
  const [isOpen, setOpen, setClose] = useBooleanState(false);
  const [uptime, setUptime] = useState<Time>(INITIAL_TIME);
  const [breakTime, setBreakTime] = useState<Time>(INITIAL_TIME);

  const result = useMemo(
    () => `
    稼働時間は${uptime.hour}時間${uptime.minute}分です。
    休憩時間は${breakTime.hour}時間${breakTime.minute}分です。
  `,
    [breakTime.hour, breakTime.minute, uptime.hour, uptime.minute],
  );

  const handleAppend = useCallback(() => {
    append({ start: '00:00', end: '00:00' });
  }, [append]);

  const handleConfirm = useCallback(
    (val: Inputs) => {
      const workTimes = val.workTimes;
      const data = getWorkingTime(workTimes);

      setUptime(data.uptime);
      setBreakTime(data.breakTime);
      setOpen();
    },
    [setOpen],
  );

  const handleReset = useCallback(() => {
    reset({
      workTimes: [
        {
          start: '09:00',
          end: '10:00',
        },
      ],
    });
  }, [reset]);

  return (
    <Component
      {...props}
      result={result}
      isDisabled={isDirty}
      isOpen={isOpen}
      setClose={setClose}
      handleAppend={handleAppend}
      handleConfirm={handleConfirm}
      handleReset={handleReset}
      handleSubmit={handleSubmit}
    />
  );
});

InputForm.displayName = 'InputForm';
