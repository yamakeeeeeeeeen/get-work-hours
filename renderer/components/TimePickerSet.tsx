import { FC, memo } from 'react';
import { UseFormMethods } from 'react-hook-form';
import { Box, Button, Typography } from '@material-ui/core';
import { Props } from '~/containers/TimePickerSet';
import { TimePicker } from '~/components/TimePicker';
import { ConfirmDialog } from '~/components/ConfirmDialog';
import { getValidation } from '~/utils/getValidation';
import { ORDINAL_NUMBERS } from '~/constants';

type ComponentProps = Omit<Props, 'fields' | 'remove'> & {
  isDisabled: boolean;
  isOpen: boolean;
  setOpen: () => void;
  setClose: () => void;
  errors: UseFormMethods['errors'];
  getValues: UseFormMethods['getValues'];
  register: UseFormMethods['register'];
  handleRemove: () => void;
};

export const Component: FC<ComponentProps> = memo(
  ({ index, value, isDisabled, isOpen, setOpen, setClose, errors, getValues, register, handleRemove }) => {
    const { biggerThanThePrevious, formatHHMM } = getValidation();

    return (
      <>
        <Box display="flex" alignItems="center" mb={3}>
          <Typography>{ORDINAL_NUMBERS[index + 1]}</Typography>
          <TimePicker
            name={`workTimes[${index}].start`}
            label="start"
            defaultValue={value.start}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('start', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.start}
            helperText={errors?.workTimes?.[index]?.start && errors?.workTimes?.[index]?.start?.message}
          />
          <Typography>~</Typography>
          <TimePicker
            name={`workTimes[${index}].end`}
            label="end"
            defaultValue={value.end}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('end', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.end}
            helperText={errors?.workTimes?.[index]?.end && errors?.workTimes?.[index]?.end?.message}
          />
          <Button variant="contained" color="secondary" disabled={isDisabled} onClick={setOpen}>
            DELETE
          </Button>
        </Box>
        <ConfirmDialog isOpen={isOpen} setClose={setClose} body="Can I delete it?" callback={handleRemove} />
      </>
    );
  },
);

Component.displayName = 'Component';
