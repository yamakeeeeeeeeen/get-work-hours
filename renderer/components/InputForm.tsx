import { FC, memo } from 'react';
import { Box, Button, createStyles, makeStyles, Theme, Typography } from '@material-ui/core';
import { Props } from '~/containers/InputForm';
import { TimePickerSet } from '~/containers/TimePickerSet';
import { ResultDialog } from '~/components/ResultDialog';
import { Inputs } from '~/pages';
import { UseFormMethods } from 'react-hook-form';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    infoText: {
      marginBottom: theme.spacing(3),
    },
  }),
);

export type ComponentProps = Omit<Props, 'append'> & {
  result: { breakTime: string; uptime: string };
  isDisabled: boolean;
  isOpen: boolean;
  setClose: () => void;
  handleAppend: () => void;
  handleConfirm: (val: Inputs) => void;
  handleReset: () => void;
  handleSubmit: UseFormMethods['handleSubmit'];
};

export const Component: FC<ComponentProps> = memo(
  ({ result, isDisabled, isOpen, setClose, handleAppend, handleConfirm, handleReset, handleSubmit, ...props }) => {
    const classes = useStyles();

    return (
      <>
        <Box maxWidth={500} mx="auto">
          <Typography className={classes.infoText}>Please enter the hours you workedüë®‚Äçüíª</Typography>
          {props.fields.map((item, index) => {
            if (item.start && item.end) {
              const value = {
                start: item.start,
                end: item.end,
              };
              return <TimePickerSet key={item.id} index={index} value={value} {...props} />;
            }
            return null;
          })}
          <Box display="flex" justifyContent="space-between">
            <Button variant="contained" color="primary" onClick={handleAppend}>
              APPEND
            </Button>
            <Button variant="contained" color="primary" onClick={handleSubmit(handleConfirm)}>
              CONFIRM
            </Button>
            <Button variant="contained" color="secondary" disabled={!isDisabled} onClick={handleReset}>
              RESET
            </Button>
          </Box>
        </Box>
        <ResultDialog isOpen={isOpen} setClose={setClose} result={result} />
      </>
    );
  },
);

Component.displayName = 'Component';
