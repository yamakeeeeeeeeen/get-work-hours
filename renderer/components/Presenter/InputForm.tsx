import { FC, memo } from 'react';
import { UseFieldArrayMethods } from 'react-hook-form';
import { Box, Button, createStyles, makeStyles, Theme, Typography } from '@material-ui/core';
import { TimePickerSet } from '~/components/Container/TimePickerSet';
import { ResultDialog } from '~/components/ResultDialog';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    infoText: {
      marginBottom: theme.spacing(3),
    },
  }),
);

type ComponentProps = {
  fields: UseFieldArrayMethods['fields'];
  remove: UseFieldArrayMethods['remove'];
  result: string;
  isOpen: boolean;
  setClose: () => void;
  handleAppend: () => void;
  handleConfirm: any;
  handleSubmit: any;
};

export const Component: FC<ComponentProps> = memo(
  ({ handleSubmit, fields, remove, result, isOpen, setClose, handleAppend, handleConfirm }) => {
    const classes = useStyles();

    return (
      <>
        <Box>
          <Typography className={classes.infoText}>着席していた時間を入力してください</Typography>
          {fields.map((item, index) => {
            const value = {
              start: item.start,
              end: item.end,
            };
            return <TimePickerSet key={item.id} index={index} value={value} fields={fields} remove={remove} />;
          })}
          <Button variant="contained" color="primary" onClick={handleAppend}>
            追加
          </Button>
          <Button variant="contained" color="primary" onClick={handleSubmit(handleConfirm)}>
            確認
          </Button>
        </Box>
        <ResultDialog isOpen={isOpen} setClose={setClose} result={result} />
      </>
    );
  },
);

Component.displayName = 'Component';
