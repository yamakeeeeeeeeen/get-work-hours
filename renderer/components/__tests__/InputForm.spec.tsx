import { Component as InputForm, ComponentProps as InputFormProps } from '~/components/InputForm';
import { snapshot } from '~/test/utils';

test('<InputForm />', () => {
  const mockFn = jest.fn();
  const props: InputFormProps = {
    fields: [],
    remove: mockFn,
    result: { breakTime: '01:00', uptime: '08:00' },
    isDisabled: true,
    isOpen: true,
    setClose: mockFn,
    handleAppend: mockFn,
    handleConfirm: mockFn,
    handleReset: mockFn,
    handleSubmit: mockFn,
  };

  snapshot(<InputForm {...props} />);
});
