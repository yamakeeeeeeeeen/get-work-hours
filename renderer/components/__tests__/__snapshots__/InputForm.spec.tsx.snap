// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputForm /> snapshot 1`] = `
<Fragment>
  <Styled(MuiBox)
    maxWidth={500}
    mx="auto"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-infoText-1"
    >
      Please enter the hours you worked👨‍💻
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      display="flex"
      justifyContent="space-between"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [
                [MockFunction] {
                  "calls": [Circular],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        variant="contained"
      >
        APPEND
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        variant="contained"
      >
        CONFIRM
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        disabled={false}
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [
                [MockFunction] {
                  "calls": [Circular],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        variant="contained"
      >
        RESET
      </WithStyles(ForwardRef(Button))>
    </Styled(MuiBox)>
  </Styled(MuiBox)>
  <ResultDialog
    isOpen={true}
    result={
      Object {
        "breakTime": "01:00",
        "uptime": "08:00",
      }
    }
    setClose={
      [MockFunction] {
        "calls": Array [
          Array [
            [MockFunction] {
              "calls": [Circular],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />
</Fragment>
`;
